src/main/java/org/dice_research/opal/launuts/lau/ExcelParser.java

```diff
@@ -0,0 +1,190 @@
+package org.dice_research.opal.launuts.lau;
+
+import org.apache.commons.collections4.ListUtils;
+import org.apache.jena.ext.com.google.common.collect.Iterables;
+import org.apache.poi.ss.usermodel.*;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.*;
+
+/**
+ * Generic Parser to parse all the sheet from the xlsx.
+ * Used instead of {@link LauCsvParser}.
+ *
+ * @author Vikrant Singh
+ */
+
+
+public class ExcelParser implements LauReaderInterface {
+
+    public static final String FILE_NAME = "EU-28-LAU-2019-NUTS-2016.xlsx";
+
+    List<LauContainer> lauContainerList = new LinkedList<LauContainer>();
+    private HashMap<String, HashMap<String, LauContainer>> index;
+
+    private Workbook workbook;
+    private Sheet sheet;
+    private Map<String, List<String>> getCodes = new HashMap<>();
+    private HashMap<String, Integer> getkeys = new HashMap<String, Integer>();
+    private HashMap<String, LauContainer> laucodeToLauContainerMap = new HashMap<String, LauContainer>();
+    private List<String> countryIds;
+    private boolean parsed = false;
+
+    @Override
+    public LauReaderInterface setLauSourceDirectory() throws IOException {
+
+        workbook = WorkbookFactory.create(new File(FILE_NAME));
+        return this;
+    }
+
+    @Override
+    public List<String> getCountryIds() throws LauReaderException {
+
+        if (!parsed)
+            parse();
+
+        return countryIds;
+    }
+
+    @Override
+    public Map<String, List<String>> getCodes(String countryId) throws LauReaderException {
+
+        if (getCountryIds().contains(countryId)) {
+            sheet = workbook.getSheet(countryId);
+
+            if (!parsed) {
+                parse();
+            }
+        }
```

Comment:
**getCountryIds() is already calling parse method**. I guess we can remove this redundant check and call or consider calling parse() from a constructor.

Reply:
This comment is common for the 1st and 3rd comment of yours, 
So the usage of having parse method is, as every function would be called independently, from the outside for fetching the data, having an extra overhead of parsing every time when the object is created is too expensive, on the other hand, the flow right now is, getcountrIds have parse so that, again and again, you don't have to parse all the countries, and same for every other method, as they work individually for different purposes.

Reply:
But all the test-cases are using **only one object**, 
[https://github.com/akumar51/LauNuts/blob/e1da26a6f94a13674fbc5fc2b75035a9f616a118/src/test/java/org/dice_research/opal/launuts/ExcelParserTest.java](url)

e.g. excelParserDE

Reply:
I have modified the class.